tests/metrics/grant_availability.py

```diff
@@ -0,0 +1,187 @@
+except KeyError:
+    message = f'{METRICS_ADDRESS_ENVVAR}, {METRICS_SIGNER_ENVVAR} and {METRICS_PASSWORD_ENVVAR}' \
+              f' are required to run grant availability metrics.'
+    raise RuntimeError(message)
+
+# Alice Configuration
+DOMAIN: str = 'ibex'
+DEFAULT_SEEDNODE_URIS: List[str] = ['https://ibex.nucypher.network:9151', ]
+ETHEREUM_PROVIDER_URI: str = 'wss://rinkeby.infura.io/ws/v3/8fb725f72a3342f6a8cf8fc7ab674123'
+INSECURE_PASSWORD: str = "METRICS_INSECURE_DEVELOPMENT_PASSWORD"
+TEMP_ALICE_DIR: str = os.path.join('/', 'tmp', 'grant-metrics')
+
+# Policy Parameters
+M: int = 1
+N: int = 1
+RATE: Wei = Wei(1)
+DURATION: datetime.timedelta = datetime.timedelta(days=1)
+
+# Tuning
+SAMPLE_RATE: int = 10  # seconds
+GAS_STRATEGY: str = 'fast'
+LABEL_PREFIX = 'random-metrics-label-'
+LABEL_SUFFIXER = lambda: os.urandom(4).hex()
+HANDPICKED_URSULA_URIS: List[str] = [
+    DEFAULT_SEEDNODE_URIS[0],  # use the seednode for granting
+]
+
+
+def make_random_bob():
+    """Generates a random ephemeral Bob instance."""
+    bob_verifying_secret = UmbralPrivateKey.gen_key()
+    bob_verifying_key = bob_verifying_secret.pubkey
+    decrypting_secret = UmbralPrivateKey.gen_key()
+    decrypting_key = decrypting_secret.pubkey
+    bob = Bob.from_public_keys(verifying_key=bob_verifying_key,
+                               encrypting_key=decrypting_key,
+                               federated_only=False)
+    return bob
+
+
+def metric_grant(alice, handpicked_ursulas: Optional[Set[Ursula]] = None) -> Policy:
+    """Perform a granting operation for metrics collection."""
+    label = f'{LABEL_PREFIX}{LABEL_SUFFIXER()}'.encode()
+    policy_end_datetime = maya.now() + DURATION
+    policy = alice.grant(m=M, n=N,
+                         handpicked_ursulas=handpicked_ursulas,
+                         expiration=policy_end_datetime,
+                         bob=make_random_bob(),
+                         label=label,
+                         rate=RATE)
+    return policy
+
+
+def collect(alice: Alice,
+            handpicked_ursulas: Optional[Set[Ursula]] = None,
+            iterations: Optional[int] = None,
+            run_forever: bool = False
+            ) -> None:
+    """Collects grant success and failure rates."""
+    policies, i, success, fail = dict(), 0, 0, 0
+    while True:
+        print(f"Attempt {i+1} of {iterations if iterations is not None else 'infinite'}")
+        start = maya.now()
+        try:
+            policy = metric_grant(alice=alice, handpicked_ursulas=handpicked_ursulas)
+        except Exception as e:
+            fail += 1
+            print(f'GRANT FAIL\n{e}')
+        else:
+            success += 1
+            policies[policy.public_key.hex()] = policy  # track
+            print(f"PEK:{policy.public_key.hex()}")
+
+        # timeit
+        end = maya.now()
+        delta = end - start
+        print(f"Completed in {(delta).total_seconds()} seconds.")
+
+        if i == iterations and not run_forever:
```

Comment:
I think we are running once more than expected and the check should be `(i+1) == iterations`. We increment `i` after we run i.e. i = 0 is really iteration 1

Reply:
Are we running on extra time here?

Lets say you want 1 iteration.

Loop 1: i = 0, perform metric grant, 0 != 1 so continue the loop, increment i to 1
Loop 2: i = 1, perform metric grant, 1 == 1 so stop.

Reply:
I knew what you meat - thanks for explanation though :-)

I think I forgot to add these two lines to my last patch.

Reply:
:heavy_check_mark: